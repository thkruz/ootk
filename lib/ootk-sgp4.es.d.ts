/**
 * @author Theodore Kruczek.
 * @description Orbital Object ToolKit (OOTK) is a collection of tools for working
 * with satellites and other orbital objects. The Sgp4 module contains a TypeScript
 * port of the 2020 version of sgp4unit.cpp from "Fundamentals of Astrodynamics and
 * Applications" by David Vallado. All of the original comments and notes are
 * inserted in the code below in order provide context to the functions and clarify
 * any adjustments made for TypeScript/JavaScript compatibility.
 *
 * @copyright MIT License.
 * Copyright (c) 2020-2021 Theodore Kruczek
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/** @interface SatelliteRecord contains all of the orbital parameters necessary for running SGP4. It is generated by Sgp4.createSatrec.*/
interface SatelliteRecord {
    a: number;
    am: number;
    alta: number;
    altp: number;
    argpdot: number;
    argpo: number;
    aycof: number;
    bstar: number;
    cc1: number;
    cc4: number;
    cc5: number;
    con41: number;
    d2: number;
    d3: number;
    d4: number;
    d5232: number;
    d5421: number;
    d5433: number;
    dedt: number;
    delmo: number;
    del1: number;
    ecco: number;
    em: number;
    epochdays: number;
    epochyr: number;
    error: number;
    eta: number;
    gsto: number;
    im: number;
    inclo: number;
    init: string;
    isimp: number;
    jdsatepoch: number;
    mdot: number;
    method: string;
    mo: number;
    mm: number;
    nddot: number;
    ndot: number;
    no: number;
    nodecf: number;
    nodedot: number;
    nodeo: number;
    om: number;
    Om: number;
    omgcof: number;
    operationmode: string;
    satnum: string;
    sinmao: number;
    t: number;
    t2cof: number;
    t3cof: number;
    t4cof: number;
    t5cof: number;
    x1mth2: number;
    x7thm1: number;
    xlcof: number;
    xmcof: number;
    xfact: number;
    xlamo: number;
    xli: number;
    xgh4: number;
    xgh3: number;
    xh2: number;
    xi2: number;
    xi3: number;
    xl2: number;
    xl3: number;
    xl4: number;
    zmol: number;
    zmos: number;
    dmdt: number;
    dnodt: number;
    domdt: number;
    e3: number;
    ee2: number;
    peo: number;
    pgho: number;
    pho: number;
    PInco: number;
    plo: number;
    se2: number;
    se3: number;
    sgh2: number;
    sgh3: number;
    sgh4: number;
    sh2: number;
    sh3: number;
    si2: number;
    si3: number;
    sl2: number;
    sl3: number;
    sl4: number;
    xgh2: number;
    xh3: number;
    tumin: number;
    radiusearthkm: number;
    irez: number;
    d3210: number;
    d3222: number;
    d4410: number;
    d4422: number;
    d5220: number;
    del2: number;
    del3: number;
    didt: number;
    atime: number;
    j2: number;
    j3: number;
    j4: number;
    mus: number;
    xke: number;
    j3oj2: number;
    xni: number;
    d2201: number;
    d2211: number;
    nm: number;
}
/** @typedef StateVector is a three dimensional position and velocity vector output from Sgp4.propagate.*/
declare type StateVector = {
    position: {
        x: number;
        y: number;
        z: number;
    } | boolean;
    velocity: {
        x: number;
        y: number;
        z: number;
    } | boolean;
};
/** @typedef vec3 is a flat three dimensional array used in vector math.*/
declare type vec3 = [number, number, number];
declare class Sgp4 {
    private static dpper;
    private static dscom;
    private static dsinit;
    private static dspace;
    private static initl;
    private static sgp4init;
    static propagate(satrec: SatelliteRecord, tsince: number): StateVector;
    private static getgravconst;
    static createSatrec(tleLine1: string, tleLine2: string, whichconst?: string, opsmode?: string): SatelliteRecord;
    static gstime(jdut1: number): number;
    static sgn(x: number): number;
    static mag(x: vec3): number;
    static cross(vec1: vec3, vec2: vec3): vec3;
    static dot(x: vec3, y: vec3): number;
    static angle(vec1: vec3, vec2: vec3): number;
    static asinh(xval: number): number;
    static newtonnu(ecc: number, nu: number): {
        e0: number;
        m: number;
    };
    static sinh(x: number): number;
    static rv2coe(r: vec3, v: vec3, mus: number): {
        p: number;
        a: number;
        ecc: number;
        incl: number;
        omega: number;
        argp: number;
        nu: number;
        m: number;
        arglat: number;
        truelon: number;
        lonper: number;
    };
    static jday(year: number, mon: number, day: number, hr: number, minute: number, sec: number): {
        jd: number;
        jdFrac: number;
    };
    static days2mdhms(year: number, days: number): {
        mon: number;
        day: number;
        hr: number;
        minute: number;
        sec: number;
    };
    static invjday(jd: number, jdfrac: number): {
        year: number;
        mon: number;
        day: number;
        hr: number;
        minute: number;
        sec: number;
    };
}
export { Sgp4 };
