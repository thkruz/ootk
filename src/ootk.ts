/**
 * @author @thkruz Theodore Kruczek
 * @description Orbital Object ToolKit (OOTK) is a collection of tools for working
 * with satellites and other orbital objects.
 *
 * @license MIT License
 * @Copyright (c) 2020-2024 Theodore Kruczek
 *
 * Many of the classes are based off of the work of @david-rc-dayton and his
 * Pious Squid library (https://github.com/david-rc-dayton/pious_squid).
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

export { Celestial } from './body/Celestial';
export { Earth } from './body/Earth';
export { Moon } from './body/Moon';
export { Sun } from './body/Sun';
export * from './objects/index';
export * from './operations/operations';
export { Sgp4 } from './sgp4/sgp4';
export * from './time/time';
export { Tle } from './tle/tle';
export * from './transforms';
export * from './types/types';
export * from './utils/constants';
export { Utils } from './utils/utils';
