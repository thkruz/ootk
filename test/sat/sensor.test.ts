/**
 * @file   Test Suite to verify sat class work as expected
 * @author Theodore Kruczek.
 * @since  1.2.0
 */

import { Sat, Sensor, SpaceObjectType } from '../../lib/ootk';

const dateObj = new Date(1661400000000);

const tle1 = '1 25544U 98067A   22203.46960946  .00003068  00000+0  61583-4 0  9996';
const tle2 = '2 25544  51.6415 161.8339 0005168  35.9781  54.7009 15.50067047350657';

const { PI } = Math;
const TAU = PI * 2; // https://tauday.com/tau-manifesto
const RAD2DEG = 360 / TAU;

describe('Basic Sensor functionality', () => {
  const sat = new Sat({ name: 'Test', tle1, tle2 });

  it('should be able to get rae coordinates', () => {
    const sensor = new Sensor({ name: 'Test', lat: 41, lon: -71, alt: 0 });

    const rae = sensor.getRae(sat, dateObj);

    rae.az *= RAD2DEG;
    rae.el *= RAD2DEG;

    expect(rae.az).toBeCloseTo(73.86522544035945);
    expect(rae.el).toBeCloseTo(-56.89115262874271);
    expect(rae.rng).toBeCloseTo(11183.206102756607);

    // Verify it works in reverse
    const rae2 = sat.getRae(sensor, dateObj);

    rae2.az *= RAD2DEG;
    rae2.el *= RAD2DEG;

    expect(rae2.az).toBeCloseTo(73.86522544035945);
    expect(rae2.el).toBeCloseTo(-56.89115262874271);
    expect(rae2.rng).toBeCloseTo(11183.206102756607);
  });

  it('should be able to determine InView without FOV', () => {
    const sensor = new Sensor({ name: 'Test', lat: 41, lon: -71, alt: 0, type: SpaceObjectType.OPTICAL });

    const inView = sensor.isSatInFov(sat, dateObj);
    const rae = sat.getRae(sensor, dateObj);

    rae.az *= RAD2DEG;
    rae.el *= RAD2DEG;
    const inView2 = sensor.isRaeInFov(rae, dateObj);

    expect(inView).toEqual(inView2);
  });

  it('should be able to determine InView with FOV', () => {
    const sensor = new Sensor({
      name: 'Test',
      lat: 41,
      lon: -71,
      alt: 0,
      minAz: 0,
      maxAz: 360,
      minEl: 0,
      maxEl: 90,
      minRng: 0,
      maxRng: 100000,
    });

    const inView = sensor.isSatInFov(sat, dateObj);
    const rae = sat.getRae(sensor, dateObj);

    rae.az *= RAD2DEG;
    rae.el *= RAD2DEG;
    const inView2 = sensor.isRaeInFov(rae, dateObj);

    expect(inView).toEqual(inView2);
  });

  it('should error if bad minAz', () => {
    const result = () => new Sensor({ name: 'Test', lat: 41, lon: -71, alt: 0, minAz: -999 });

    expect(result).toThrow();
  });

  it('should error if bad maxAz', () => {
    const result = () => new Sensor({ name: 'Test', lat: 41, lon: -71, alt: 0, maxAz: -999 });

    expect(result).toThrow();
  });

  it('should error if bad minEl', () => {
    const result = () => new Sensor({ name: 'Test', lat: 41, lon: -71, alt: 0, minEl: -999 });

    expect(result).toThrow();
  });

  it('should error if bad maxEl', () => {
    const result = () => new Sensor({ name: 'Test', lat: 41, lon: -71, alt: 0, maxEl: -999 });

    expect(result).toThrow();
  });

  it('should error if bad minRng', () => {
    const result = () => new Sensor({ name: 'Test', lat: 41, lon: -71, alt: 0, minRng: -999 });

    expect(result).toThrow();
  });
  it('should error if bad maxRng', () => {
    const result = () => new Sensor({ name: 'Test', lat: 41, lon: -71, alt: 0, maxRng: -999 });

    expect(result).toThrow();
  });

  it('should error if bad lat', () => {
    const result = () => new Sensor({ name: 'Test', lat: -999, lon: -71, alt: 0 });

    expect(result).toThrow();
  });

  it('should error if bad lon', () => {
    const result = () => new Sensor({ name: 'Test', lat: 41, lon: -999, alt: 0 });

    expect(result).toThrow();
  });

  it('should error if bad alt', () => {
    const result = () => new Sensor({ name: 'Test', lat: 41, lon: -71, alt: -999 });

    expect(result).toThrow();
  });

  it('should create a lookangles array', () => {
    const sensor = new Sensor({
      name: 'Test',
      lat: 41,
      lon: -71,
      alt: 0,
      minAz: 0,
      maxAz: 360,
      minEl: 0,
      maxEl: 90,
      minRng: 0,
      maxRng: 100000,
    });

    const lookangles = sensor.calculatePasses(1440 * 60, sat, dateObj);

    expect(lookangles).toMatchSnapshot();
  });
});
