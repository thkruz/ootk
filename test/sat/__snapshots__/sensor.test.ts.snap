// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Sensor functionality should create a lookangles array 1`] = `
Array [
  Object {
    "az": 185.0321047837688,
    "el": 0.018932126219342378,
    "rng": 2320.7003088160936,
    "time": 2022-08-25T07:51:46.000Z,
    "type": 0,
  },
  Object {
    "az": 286.6828392807201,
    "el": -0.03143720527465417,
    "maxElPass": 8.887905466445714,
    "rng": 2346.1443165688643,
    "time": 2022-08-25T08:00:03.000Z,
    "type": 2,
  },
  Object {
    "az": 126.07032765927691,
    "el": 0.011464577747703686,
    "rng": 2327.848319797019,
    "time": 2022-08-25T09:27:12.000Z,
    "type": 0,
  },
  Object {
    "az": 338.51245540232674,
    "el": -0.004776656652769985,
    "maxElPass": 28.303560760507143,
    "rng": 2353.0795991791333,
    "time": 2022-08-25T09:37:34.000Z,
    "type": 2,
  },
  Object {
    "az": 24.30450924794974,
    "el": 0.021759132371179508,
    "rng": 2365.173163410431,
    "time": 2022-08-25T19:22:17.000Z,
    "type": 0,
  },
  Object {
    "az": 230.35934291906204,
    "el": -0.029094115923272833,
    "maxElPass": 34.603325622933625,
    "rng": 2350.6620737684816,
    "time": 2022-08-25T19:32:54.000Z,
    "type": 2,
  },
  Object {
    "az": 77.30327662230586,
    "el": 0.009671715968386793,
    "rng": 2358.781639064529,
    "time": 2022-08-25T21:00:08.000Z,
    "type": 0,
  },
  Object {
    "az": 171.14122538519032,
    "el": -0.04352163512801034,
    "maxElPass": 7.334113831175358,
    "rng": 2346.370746417976,
    "time": 2022-08-25T21:08:00.000Z,
    "type": 2,
  },
]
`;
